Add an eslint rule for checking interactivity of tooltip trigger

broccolinisoup opened the pull request:
Adding an eslint rule to check if the `Tooltip` component includes an interactive trigger. 

## Rule Details

This rule enforces to use interactive elements as tooltip triggers. Interactive elements can be Primer `Button`, `IconButton` and `Link` components or native elements like `button`, `a` with an `href` attribute, `select`, `textarea`, `summary` and `input` (that is not a `hidden` type).

👎 Examples of **incorrect** code for this rule:

```jsx
/* eslint primer-react/a11y-tooltip-non-interactive-trigger: "error" */
import {Tooltip} from '@primer/react'
const App = () => (
  <Tooltip text="Tooltip text">
    <div>Tooltip trigger</div>
  </Tooltip>
)
```

👍 Examples of **correct** code for this rule:

```jsx
/* eslint primer-react/a11y-tooltip-non-interactive-trigger: "error" */
import {Tooltip, Button} from '@primer/react'
const App = () => (
  <Tooltip text="Supplementary text" type="description">
    <Button
      onClick={() => {
        /* do something */
      }}
    >
      Save
    </Button>
  </Tooltip>
)
```

## Options

- `skipImportCheck` (default: `false`)

  By default, the `a11y-tooltip-non-interactive-trigger` rule will only check for interactive elements in components that are imported from `@primer/react`. You can disable this behavior by setting `skipImportCheck` to `true`. This is used for internal linting in the [primer/react](https://github.com/prime/react) repository.

broccolinisoup commented on src/rules/non-interactive-tooltip-trigger.js:
```diff
@@ -0,0 +1,126 @@
+      JSXElement(jsxNode) {
+        // If `skipImportCheck` is true, this rule will check for direct slot children
+        // in any components (not just ones that are imported from `@primer/react`).
+        const skipImportCheck = context.options[0] ? context.options[0].skipImportCheck : false
```
I am unsure what this `skipImportCheck` does. Do I need it here or not necessary maybe in this rule? 👀

joshblack replied:
@broccolinisoup one guess (I'm also not sure) is that we need a way to verify that the `<Tooltip>` component comes from the `@primer/react` package and is not a local `Tooltip` for the project. Skipping it may be useful in `primer/react` since we can assume that `Tooltip` is the correct `Tooltip` and the import check would fail since we don't import from `@primer/react`.

broccolinisoup replied:
Oh I see! That makes sense now. Thank you!

broccolinisoup commented on src/rules/non-interactive-tooltip-trigger.js:
```diff
@@ -0,0 +1,126 @@
+        .filter(gChild => gChild.type === 'JSXElement')
+        .some(gChild => {
+          const gChildName = getJSXOpeningElementName(gChild.openingElement)
+          // TODO: How can I check all child nodes?
```
I am trying to understand if I should check grandchildrens' interactivity? 🤔 And is so, is there a way to access all descendants here?

joshblack replied:
I definitely think it can be helpful to recursively check children but unfortunately we're limited to the owner tree and won't be able to see the render result of children 😞 Definitely worth the trade-off though, we just may error out incorrectly if a component renders something interactive.

One quick way to get a flat list of all the jsxElement children could be a pre-order traversal:

```jsx
function getChildren(node) {
  if (Array.isArray(node.children)) {
    return node.children.filter((child) => {
      return child.type === 'JSXElement';
    }).flatMap((child) => {
      return [child, ...getChildren(child)];
    });
  }
  return [];
}
```

Hope this works 😅

broccolinisoup replied:
>I definitely think it can be helpful to recursively check children but unfortunately we're limited to the owner tree and won't be able to see the render result of children 😞

Yeah, it makes things harder but that is okay. The main reason I tried recursive, I wanted to get all children as a flat list and that function you wrote works perfectly for this ✨

joshblack commented on src/rules/non-interactive-tooltip-trigger.js:
```diff
@@ -0,0 +1,126 @@
+      nonInteractiveTrigger:
+        'The `Tooltip` component expects a single React element that contains interactive content. Consider using a `<button>` or equivalent interactive element instead.',
+      anchorTagWithoutHref:
+        'Anchor tags without an href attribute are not be considered interactive, therefore can not be used as a trigger for a tooltip.',
```
```suggestion
        'Anchor tags without an href attribute are not interactive, therefore they cannot be used as a trigger for a tooltip. Please add an href attribute or use an alternative interactive element instead',
```

joshblack commented on src/rules/non-interactive-tooltip-trigger.js:
```diff
@@ -0,0 +1,126 @@
+      anchorTagWithoutHref:
+        'Anchor tags without an href attribute are not be considered interactive, therefore can not be used as a trigger for a tooltip.',
+      hiddenInput:
+        'Hidden inputs are not be considered interactive, therefore can not be used as a trigger for a tooltip.',
```
```suggestion
        'Hidden inputs are not interactive and cannot be used as a trigger for a tooltip. Please use an alternate input type or use a different interactive element instead',
```

joshblack commented on src/rules/non-interactive-tooltip-trigger.js:
```diff
@@ -0,0 +1,126 @@
+  return !hasHiddenType
+}
+
+const checkTriggerElement = jsxNode => {
```
One idea for collecting errors and going through all children here could be a block like:

```suggestion
const checks = [
  {
    id: 'anchorTagWithoutHref',
    filter: (jsxElement) => {
      return jsxElement.openingElement.name === 'a';
    },
    check: isInteractiveAnchor,
  },
  {
    id: 'hiddenInput',
    filter: (jsxElement) => {
      return jsxElement.openingElement.name === 'input';
    },
    check: isInteractiveInput,
  },
];

const checkTriggerElement = jsxNode => {  
  const elements = [jsxNode, ...getAllChildren(jsxNode)];
  const hasInteractiveElement = elements.find(element => {
    return isInteractive(element);
  });
  
  if (hasInteractiveElement) {
    return;
  }
  
  const errors = new Set();
  
  for (const element of elements) {
    for (const check of checks) {
      if (!check.filter(element)) {
        continue;
      }
      
      if (!check.check(element)) {
        errors.add(check.id);
      }
    }
  }
};
```

Basically get all the elements in this part of the JSX tree, check to see if there is an interactive one, and if not try and find the relevant errors and report whatever the first one is.

broccolinisoup replied:
I like this 💖 It is cleaner. I had to play with it to cover all test cases but it helped a lot.

joshblack reviewed (commented):
So cool! 🥳 Loving this plugin. Left a couple of comments and questions, let me know what you think 👀

broccolinisoup commented on src/rules/non-interactive-tooltip-trigger.js:
```diff
@@ -0,0 +1,126 @@
+const {getJSXOpeningElementName} = require('../utils/get-jsx-opening-element-name')
+const {getJSXOpeningElementAttribute} = require('../utils/get-jsx-opening-element-attribute')
+
+isInteractive = child => {
```
Ah sorry for my ignorance! Corrected it.

broccolinisoup commented on src/rules/__tests__/non-interactive-tooltip-trigger.test.js:
```diff
@@ -0,0 +1,81 @@
+      ]
+    },
+    {
+      code: `import {Tooltip, Button} from '@primer/react';<Tooltip aria-label="Supplementary text" direction="e"><heading><span>Save</span></heading></Tooltip>`,
```
🙈 I meant to write `header` .I know it is not a very practical test case but wanted to add more extreme examples. Let me know if you have any feedback on this!

joshblack replied:
Makes sense!

broccolinisoup commented on src/rules/__tests__/non-interactive-tooltip-trigger.test.js:
```diff
@@ -0,0 +1,81 @@
+  ],
+  invalid: [
+    {
+      code: `import {Tooltip} from '@primer/react';<Tooltip type="description" text="supportive text" direction="e"><button>save</button><button>cancel</button></Tooltip>`,
```
I know 😞 I feel the same way but funny thing is when I start adding the spaces, it returns the "singleChild" error which I think it takes spaces as children as well. I am a bit concerned how this is going to work out in the actual code base but looking at other tests, they don't seem to have spaces as well maybe because of this? 🤷🏻‍♀️

joshblack replied:
@broccolinisoup if you filter out the elements which are not JSXElement's do you get the same single child error? There may be some `Literal`'s sneaking in here (I hope lol)

broccolinisoup replied:
No, it worked 🤩 ([commit](https://github.com/primer/eslint-plugin-primer-react/pull/51/commits/a65aa32c612c7fe952ec47bb3d926cf8adae9fab#diff-a54d2be13ca1d027a6df1f4788ffdc6b2e57127aa47ce8bf42dcff3bcca302fdR162)) I reformatted all now 💖

joshblack reviewed (approved):
So good! 🔥

broccolinisoup commented:
@joshblack I realised that I forgot to add some Primer components 🙈  Like `Link` and `TextInput`, so I added them along with docs. I also renamed the file to have prefix `a11y-tooltip` for accessibility team to use filter the disabled eslint rules for accessibility score cards. All in [this commit](https://github.com/primer/eslint-plugin-primer-react/pull/51/commits/a65aa32c612c7fe952ec47bb3d926cf8adae9fab#diff-a54d2be13ca1d027a6df1f4788ffdc6b2e57127aa47ce8bf42dcff3bcca302fdR162) Let me know if you have any feedback! Thanks so much again for the review 🙏🏻

colebemis commented on src/rules/a11y-tooltip-non-interactive-trigger.js:
```diff
@@ -0,0 +1,184 @@
+    docs: {
+      url:
+        'https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/tree/HEAD/docs/rules/no-noninteractive-element-interactions.md',
+      description: 'Non-interactive elements should not be assigned mouse or keyboard event listeners.'
```
Should this description be more specific to Tooltip?

broccolinisoup replied:
Oh I think I copied this from somewhere to change it later and forgot! Thanks for the catch. I'll update it

broccolinisoup replied:
I ended up removing this because I don't see it in other rules and I don't also know where it would appear 🤔 Let me know what you think

colebemis commented:
@broccolinisoup You'll also need to add this rule to the `recommended` file: https://github.com/primer/eslint-plugin-primer-react/blob/main/src/configs/recommended.js

broccolinisoup commented on src/rules/__tests__/a11y-tooltip-non-interactive-trigger.test.js:
```diff
@@ -0,0 +1,167 @@
+    </Tooltip>`,
+
+    `import {Tooltip, Button} from '@primer/react';
+    <Tooltip aria-label="Supplementary text" direction="e">
+      <div>
+        <Button>Save</Button>
+      </div>
+    </Tooltip>`,
```
Thanks for raising this! I reached out to @TylerJDev to discuss about it. 

In the current implementation, this is an allowed case.  Even though `div` here is non-interactive but it has `Button` as an interactive element and as long as the tooltip element is adjacent to the trigger element in the dom that should be fine. This was our thinking and the implementation. I am trying to understand if we need a hard rule on that to disallow being interactive element as nested.

Do you have any immediate concerns that comes to mind?

colebemis reviewed (approved):
Looks great, @broccolinisoup! Just left one naming suggestion.

colebemis reviewed (approved):
Looks great, @broccolinisoup! Just left one naming suggestion.

broccolinisoup commented on src/configs/recommended.js:
```diff
@@ -10,7 +10,8 @@ module.exports = {
     'primer-react/no-deprecated-colors': 'warn',
-    'primer-react/no-system-props': 'warn'
+    'primer-react/no-system-props': 'warn',
+    'primer-react/a11y-tooltip-non-interactive-trigger': 'error'
```
I love the `tooltip-interactive-trigger` Do you have any concern keeping the `a11y`? I added this to help accessibility team to scan easily for scorecards. I think they leverage this prefix to scan the disabled rules. https://github.com/search?q=aria-label-is-well-formatted+AND+repo%3Agithub%2Fgithub+&type=code

broccolinisoup replied:
I can do `a11y-tooltip-interactive-trigger` ?

colebemis replied:
Works for me! 👍
